tck/src/main/java/org/eclipse/microprofile/lra/tck/participant/nonjaxrs/valid/ValidLRACSParticipant.java

```diff
@@ -0,0 +1,160 @@
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.eclipse.microprofile.lra.tck.participant.api.ParticipatingTckResource.ACCEPT_PATH;
+import static org.eclipse.microprofile.lra.tck.participant.api.ParticipatingTckResource.RECOVERY_PARAM;
+import static org.eclipse.microprofile.lra.tck.participant.nonjaxrs.valid.ValidLRAParticipant.COMPENSATED_COUNT_PATH;
+import static org.eclipse.microprofile.lra.tck.participant.nonjaxrs.valid.ValidLRAParticipant.COMPLETED_COUNT_PATH;
+import static org.eclipse.microprofile.lra.tck.participant.nonjaxrs.valid.ValidLRAParticipant.STATUS_COUNT_PATH;
+
+/**
+ * Valid participant resource containing async non-JAX-RS participant methods with 
+ * {@link CompletionStage} return types
+ */
+@ApplicationScoped
+@Path(ValidLRACSParticipant.ROOT_PATH)
+public class ValidLRACSParticipant {
+    
+    public static final String ROOT_PATH = "valid-cs-participant1";
+    public static final String ENLIST_WITH_COMPLETE = "enlist-complete";
+    public static final String ENLIST_WITH_COMPENSATE = "enlist-compensate";
+
+    private final AtomicInteger completedCount = new AtomicInteger(0);
+    private final AtomicInteger compensatedCount = new AtomicInteger(0);
+    private final AtomicInteger statusCount = new AtomicInteger(0);
+    
+    private int recoveryPasses;
+
+    @GET
+    @Path(ENLIST_WITH_COMPLETE)
+    @LRA(value = LRA.Type.REQUIRED)
+    public Response enlistWithComplete(@HeaderParam(LRA.LRA_HTTP_CONTEXT_HEADER) String lraId) {
+        return Response.ok().build();
+    }
+    
+    @GET
+    @Path(ENLIST_WITH_COMPENSATE)
+    @LRA(value = LRA.Type.REQUIRED, cancelOn = Response.Status.INTERNAL_SERVER_ERROR)
+    public Response enlistWithCompensate(@HeaderParam(LRA.LRA_HTTP_CONTEXT_HEADER) String lraId) {
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+    }
+
+    @Compensate
+    public CompletionStage<Void> compensate(URI lraId) {
+        assert lraId != null;
+        
+        return CompletableFuture.runAsync(() -> {
+            compensatedCount.incrementAndGet();
+            
+            simulateLongRunningCompensation();
+        });
+    }
+
+    @Complete
+    public CompletionStage<Response> complete(URI lraId) {
+        assert lraId != null;
+        
+        return CompletableFuture.supplyAsync(() -> {
+            completedCount.incrementAndGet();
+            
+            simulateLongRunningCompensation();
+            return Response.accepted().build(); // Completing
```

Comment:
Even though this is just a TCK test it is not recommended to use futures in this way - the developer should instead just return a status without futures. If you do include the method then your test should validate that the implementation also calls the `@Status` method.

Reply:
That is exactly the point of the test, to verify that this not supposed behaviour is still respected. The test is here [1] and it verifies that if Complete with CompletionStage<Response> returns status code ACCEPTED (meaning Completing) then non JAX-RS Status method is called which return CompletionStage<Participant>. The test is testing both CompletionStage return types and that the behavior of immediate participant states is respected in non JAX-RS methods with CSs. 

[1] https://github.com/eclipse/microprofile-lra/pull/148/files#diff-c7998ff70fcd282ad67a4e1601f52093R147

Reply:
Ah yes, very good, I missed the check at the end of the test. Thanks